hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/metrics/UserMetricsContext.java

```diff
@@ -0,0 +1,165 @@
+        return getMetric(name, Unit.COUNT);
+    }
+
+    Gauge getGauge(String name, Unit unit) {
+        return getMetric(name, unit);
+    }
+
+    public void collect(MetricTagger tagger, MetricsCollectionContext context) {
+        if (onlyMetric != null) {
+            context.collect(tagger, onlyName, ProbeLevel.INFO, toProbeUnit(onlyUnit), onlyMetric.get());
+        } else if (metrics != null) {
+            metrics.forEach((name, metric) ->
+                    context.collect(tagger, name, ProbeLevel.INFO, toProbeUnit(metric.unit()), metric.get()));
+        }
+    }
+
+    private Metric getMetric(String name, Unit unit) {
+        Metric res = null;
+        if (name.equals(onlyName)) {
+            res = onlyMetric;
+        } else if (metrics != null) {
+            res = metrics.get(name);
+        }
+        // register metric on first use
+        if (res == null) {
+            res = new MetricImpl(name, unit);
+            if (onlyName == null) {
+                // the first and so far the only metric
+                onlyName = name;
+                onlyMetric = res;
+                onlyUnit = unit;
+            } else {
+                // 2 or more metrics
+                if (metrics == null) {
+                    metrics = new HashMap<>();
+                    metrics.put(onlyName, onlyMetric);
+
+                    onlyName = null;
+                    onlyMetric = null;
+                    onlyUnit = null;
+                }
+                metrics.put(name, res);
+            }
+        }
+        return res;
+    }
+
+    private ProbeUnit toProbeUnit(Unit unit) {
+        switch (unit) {
+            case COUNT:
+                return ProbeUnit.COUNT;
+            case MS:
+                return ProbeUnit.MS;
+            case BYTES:
+                return ProbeUnit.BYTES;
+            default:
+                throw new RuntimeException("Unhandled metrics unit " + unit);
+        }
+    }
+
+    private interface Metric extends Counter, Gauge {
+
+        long get();
+
+        Unit unit();
+
+    }
+
+    private static final class MetricImpl implements Metric {
+
+        private static final AtomicLongFieldUpdater<MetricImpl> VOLATILE_VALUE_UPDATER =
+                AtomicLongFieldUpdater.newUpdater(MetricImpl.class, "value");
+
+        private final String name;
+        private final Unit unit;
+        private volatile long value;
+
+        MetricImpl(String name, Unit unit) {
+            this.name = name;
+            this.unit = unit;
+        }
+
+        @Nonnull @Override
+        public String name() {
+            return name;
+        }
+
+        @Override
+        public Unit unit() {
+            return unit;
+        }
+
+        @Override
+        public void set(long newValue) {
+            VOLATILE_VALUE_UPDATER.lazySet(this, newValue);
+        }
+
+        @Override
+        public void inc() {
+            VOLATILE_VALUE_UPDATER.lazySet(this, value + 1);
```

Comment:
this approach means that counter can only be incremented by a single thread (like in mapUsingContextAsync). This should be documented, in this case.

Reply:
Even if we document it, we still won't have a workaround for the ForkJoinPool case... Should we maybe give up on `lazySet`?

Reply:
We could have a different thread-safe counter, perhaps.